src/Numerics/SoaSphericalTensor.h

```diff
@@ -152,21 +151,68 @@ class SoaSphericalTensor
     const size_t nR     = nElec * Npbc; // total number of positions to evaluate
     const size_t offset = Nlm * nR;     // stride for v/gx/gy/gz/l
 
-    auto* xyz_ptr          = xyz.data();
-    auto* Ylm_vgl_ptr      = Ylm_vgl.data();
-    auto* factorLM__ptr    = factorLM_.data();
-    auto* factorL__ptr     = factorL_.data();
-    auto* factor2L__ptr    = factor2L_.data();
-    auto* norm_factor__ptr = norm_factor_.data();
+    auto* xyz_ptr          = xyz.device_data();
+    auto* Ylm_vgl_ptr      = Ylm_vgl.device_data();
+    auto* factorLM__ptr    = factorLM_.device_data();
+    auto* factorL__ptr     = factorL_.device_data();
+    auto* factor2L__ptr    = factor2L_.device_data();
+    auto* norm_factor__ptr = norm_factor_.device_data();
 
     PRAGMA_OFFLOAD("omp target teams distribute parallel for \
-                    map(to:factorLM__ptr[:Nlm], factorL__ptr[:Lmax+1], norm_factor__ptr[:Nlm], factor2L__ptr[:Lmax+1]) \
-                    map(to: xyz_ptr[:nR*3], Ylm_vgl_ptr[:5*nR*Nlm])")
+                    is_device_ptr(factorLM__ptr, factorL__ptr, norm_factor__ptr, factor2L__ptr) \
+                    is_device_ptr(xyz_ptr, Ylm_vgl_ptr)")
     for (uint32_t ir = 0; ir < nR; ir++)
       evaluateVGL_impl(xyz_ptr[0 + 3 * ir], xyz_ptr[1 + 3 * ir], xyz_ptr[2 + 3 * ir], Ylm_vgl_ptr + (ir * Nlm), Lmax,
                        factorL__ptr, factorLM__ptr, factor2L__ptr, norm_factor__ptr, offset);
   }
 
+ /** Multi‐center version (flattened). 
+ *  xyz shape = [nElec * num_centers, nPBC, 3]
+ *  Ylm_vgl shape = [5, (nElec * num_centers), nPBC, nYlm]
+ *
+ */
+inline void batched_evaluateVGL_multiCenter(const OffloadArray3D& xyz, OffloadArray4D& Ylm_vgl, size_t num_centers) const
+{
+    const size_t nElec = xyz.size(0);
+    const size_t Npbc_per_center = xyz.size(1) / num_centers; // Total images = Npbc * num_centers
+    assert(xyz.size(2) == 3);
+    assert(Ylm_vgl.size(0) == 5);
+    assert(Ylm_vgl.size(1) == nElec);
+    assert(Ylm_vgl.size(2) == xyz.size(1));
+    const size_t Nlm = Ylm_vgl.size(3);
+    
+    const size_t total_positions = nElec * Npbc_per_center * num_centers;
+    const size_t offset = Nlm * total_positions;
+
+    auto* xyz_ptr = xyz.device_data();
+    auto* Ylm_vgl_ptr = Ylm_vgl.device_data();
+    auto* factorLM_ptr = factorLM_.device_data();
+    auto* factorL_ptr = factorL_.device_data();
+    auto* factor2L_ptr = factor2L_.device_data();
+    auto* norm_factor_ptr = norm_factor_.device_data();
+
+    PRAGMA_OFFLOAD("omp target teams distribute parallel for collapse(3) \
+                    is_device_ptr(factorLM_ptr, factorL_ptr, norm_factor_ptr, factor2L_ptr, xyz_ptr, Ylm_vgl_ptr)")
+    for (size_t i_c = 0; i_c < num_centers; ++i_c) {
```

Comment:
is this actually doing anything differently from the existing function?
The loop treats every image/center/elec individually (in the existing function, this is clearer because we don't nest the loops, but it looks to me like this is doing the same thing)

Reply:
also, it's not clear to me that the labels here are correct; it's a dense rectangular loop, so the result will still be correct, but if centers and images are in the same combined dimension, then no matter how they are ordered, they should come after the nElec loop (again, it doesn't actually do anything incorrect because of the dense packing, but the actual data doesn't seem to correspond to the indices as used here, so if we ever want to do anything that relies on having those correct, then this will be very confusing for someone looking at it in the future)